hal/src/main/native/athena/PDP.cpp

```diff
@@ -62,77 +141,200 @@ HAL_Bool HAL_CheckPDPChannel(int32_t channel) {
+        break;
+      case 14:
+        raw = (static_cast<uint32_t>(pdpStatus.bits.chan15_h4) << 6) |
+              pdpStatus.bits.chan15_l6;
+        break;
+      case 15:
+        raw = (static_cast<uint32_t>(pdpStatus.bits.chan16_h2) << 8) |
+              pdpStatus.bits.chan16_l8;
+        break;
+    }
+  }
 
-  return current;
+  return raw * 0.125;
 }
 
-double HAL_GetPDPTotalCurrent(int32_t module, int32_t* status) {
-  if (!checkPDPInit(module, status)) return 0;
+double HAL_GetPDPTotalCurrent(HAL_PDPHandle handle, int32_t* status) {
+  PdpStatusEnergy pdpStatus;
+  int32_t length = 0;
+  uint64_t receivedTimestamp = 0;
 
-  double current;
+  HAL_ReadCANPeriodicPacket(handle, StatusEnergy, pdpStatus.data, &length,
+                            &receivedTimestamp, TimeoutMs, StatusPeriodMs,
+                            status);
 
-  *status = pdp[module]->GetTotalCurrent(current);
-
-  return current;
+  uint32_t raw;
+  raw = pdpStatus.bits.TotalCurrent_125mAperunit_h8;
+  raw <<= 4;
+  raw |= pdpStatus.bits.TotalCurrent_125mAperunit_l4;
+  return 0.125 * raw;
 }
 
-double HAL_GetPDPTotalPower(int32_t module, int32_t* status) {
-  if (!checkPDPInit(module, status)) return 0;
-
-  double power;
-
-  *status = pdp[module]->GetTotalPower(power);
-
-  return power;
+double HAL_GetPDPTotalPower(HAL_PDPHandle handle, int32_t* status) {
+  PdpStatusEnergy pdpStatus;
+  int32_t length = 0;
+  uint64_t receivedTimestamp = 0;
+
+  HAL_ReadCANPeriodicPacket(handle, StatusEnergy, pdpStatus.data, &length,
+                            &receivedTimestamp, TimeoutMs, StatusPeriodMs,
+                            status);
+
+  uint32_t raw;
+  raw = pdpStatus.bits.Power_125mWperunit_h4;
+  raw <<= 8;
+  raw |= pdpStatus.bits.Power_125mWperunit_m8;
+  raw <<= 4;
+  raw |= pdpStatus.bits.Power_125mWperunit_l4;
+  return 0.125 * raw;
 }
 
-double HAL_GetPDPTotalEnergy(int32_t module, int32_t* status) {
-  if (!checkPDPInit(module, status)) return 0;
-
-  double energy;
-
-  *status = pdp[module]->GetTotalEnergy(energy);
-
-  return energy;
+double HAL_GetPDPTotalEnergy(HAL_PDPHandle handle, int32_t* status) {
+  PdpStatusEnergy pdpStatus;
+  int32_t length = 0;
+  uint64_t receivedTimestamp = 0;
+
+  HAL_ReadCANPeriodicPacket(handle, StatusEnergy, pdpStatus.data, &length,
+                            &receivedTimestamp, TimeoutMs, StatusPeriodMs,
+                            status);
+
+  uint32_t raw;
+  raw = pdpStatus.bits.Energy_125mWPerUnitXTmeas_h4;
+  raw <<= 8;
+  raw |= pdpStatus.bits.Energy_125mWPerUnitXTmeas_mh8;
+  raw <<= 8;
+  raw |= pdpStatus.bits.Energy_125mWPerUnitXTmeas_ml8;
+  raw <<= 8;
+  raw |= pdpStatus.bits.Energy_125mWPerUnitXTmeas_l8;
+
+  double energyJoules = 0.125 * raw; /* mW integrated every TmeasMs */
+  energyJoules *= 0.001;             /* convert from mW to W */
+  energyJoules *=
+      pdpStatus.bits
+          .TmeasMs_likelywillbe20ms_; /* multiplied by TmeasMs = joules */
+  return 0.125 * raw;
```

Comment:
We should add the CTRE comment here about this magic number: `/* 7.3 fixed pt value in Amps */`

Reply:
Thats not what this one is...